File: PitBoard(Iphone)/PitBoard/UI/Events/EventsFilterPanel.swift
import SwiftUI
import PitBoardKit

/// Tipos de sesión que se pueden elegir en el filtro rápido — se deja fuera `.other`
/// ("" — sin clasificar), que no es algo que nadie elija filtrar a propósito. Equivalente
/// exacto de `SESSION_TYPE_FILTER_OPTIONS` en `EventsScreen.kt`.
private let sessionTypeFilterOptions: [SessionBadgeType] = [.race, .qualy, .sprint, .practice]

/// Panel de filtro de Eventos (búsqueda + series + tipo de sesión) — vive detrás del botón
/// de embudo de la barra superior, ver `EventsScreen`. Extraído de `EventsScreen.swift`
/// (Fase 4 del diagnóstico) sin cambiar ningún comportamiento, solo para que ese archivo
/// deje de concentrar navegación + filtros + lista + diálogos en un único sitio de 620
/// líneas.
struct EventsFilterPanel: View {
    @Binding var searchQuery: String
    let selectedSeries: Set<RaceSeries>
    let selectedSessionTypes: Set<String>
    let seriesConfigByKey: [RaceSeries: SeriesConfigModel]
    let onSeriesChange: (Set<RaceSeries>) -> Void
    let onSessionTypesChange: (Set<String>) -> Void

    @Environment(AppSettingsRepository.self) private var settings
    @Environment(\.pitBoardColors) private var colors

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Image(systemName: "magnifyingglass").foregroundStyle(.secondary).accessibilityHidden(true)
                TextField(settings.t("events_search_placeholder"), text: $searchQuery)
                    .accessibilityIdentifier("events.searchField")
                if !searchQuery.isEmpty {
                    Button { searchQuery = "" } label: { Image(systemName: "xmark.circle.fill") }
                        .foregroundStyle(.secondary)
                        .accessibilityLabel(settings.t("cd_clear_search"))
                }
            }
            .padding(10)
            .background(colors.surfaceVariant.opacity(0.4))
            .clipShape(RoundedRectangle(cornerRadius: PitBoardShapes.small))

            FlowLayout(spacing: 8) {
                FilterChipView(label: settings.t("events_filter_all_series"), selected: selectedSeries.isEmpty) {
                    onSeriesChange([])
                }
                ForEach(RaceSeries.allCases) { series in
                    let active = selectedSeries.contains(series)
                    FilterChipView(label: seriesConfigByKey[series]?.tag ?? series.defaultTag, selected: active) {
                        var updated = selectedSeries
                        if active { updated.remove(series) } else { updated.insert(series) }
                        onSeriesChange(updated)
                    }
                }
            }

            FlowLayout(spacing: 8) {
                FilterChipView(label: settings.t("events_filter_all_sessions"), selected: selectedSessionTypes.isEmpty) {
                    onSessionTypesChange([])
                }
                ForEach(sessionTypeFilterOptions, id: \.rawValue) { badge in
                    let active = selectedSessionTypes.contains(badge.rawValue)
                    FilterChipView(label: settings.t(badge.labelKey), selected: active) {
                        var updated = selectedSessionTypes
                        if active { updated.remove(badge.rawValue) } else { updated.insert(badge.rawValue) }
                        onSessionTypesChange(updated)
                    }
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
    }
}

private struct FilterChipView: View {
    let label: String
    let selected: Bool
    let action: () -> Void
    @Environment(\.pitBoardColors) private var colors

    var body: some View {
        Button(action: action) {
            Text(label)
                .font(.footnote.weight(.medium))
                .padding(.horizontal, 14)
                .padding(.vertical, 8)
                .background(selected ? colors.primaryContainer : colors.surfaceVariant.opacity(0.5))
                .foregroundStyle(selected ? colors.onPrimaryContainer : colors.onSurfaceVariant)
                .clipShape(Capsule())
        }
        .buttonStyle(.plain)
        // 05/09/2026 (Fase 2, accesibilidad): sin esto VoiceOver nunca anunciaba si un chip
        // de filtro estaba activo o no.
        .accessibilityAddTraits(selected ? .isSelected : [])
    }
}

/// Envuelve chips en varias líneas, ajustando al ancho disponible — equivalente de
/// `FlowRow` de Compose (SwiftUI no trae un layout de ajuste de línea nativo antes de
/// iOS 16 `Layout`; con iOS 17 como mínimo del proyecto, se implementa directamente con
/// el protocolo `Layout`).
private struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        var totalHeight: CGFloat = 0
        var lineWidth: CGFloat = 0
        var lineHeight: CGFloat = 0

        for subview in subviews {
            let size = subview.sizeThatFits(.unspecified)
            if lineWidth > 0, lineWidth + size.width > maxWidth {
                totalHeight += lineHeight + spacing
                lineWidth = 0
                lineHeight = 0
            }
            lineWidth += size.width + spacing
            lineHeight = max(lineHeight, size.height)
        }
        totalHeight += lineHeight
        return CGSize(width: maxWidth.isFinite ? maxWidth : lineWidth, height: totalHeight)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x = bounds.minX
        var y = bounds.minY
        var lineHeight: CGFloat = 0

        for subview in subviews {
            let size = subview.sizeThatFits(.unspecified)
            if x > bounds.minX, x + size.width > bounds.maxX {
                x = bounds.minX
                y += lineHeight + spacing
                lineHeight = 0
            }
            subview.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(size))
            x += size.width + spacing
            lineHeight = max(lineHeight, size.height)
        }
    }
}

File: PitBoard(Android)/app/src/main/java/com/pitboard/app/ui/EventsFilterPanel.kt
package com.pitboard.app.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.pitboard.app.data.RaceSeries
import com.pitboard.app.data.SeriesConfigEntity
import com.pitboard.app.data.SessionBadgeType
import com.pitboard.app.i18n.tr

/** Tipos de sesión que se pueden elegir en el filtro rápido — se deja fuera OTHER ("" — sin
 *  clasificar), que no es algo que nadie elija filtrar a propósito. */
private val SESSION_TYPE_FILTER_OPTIONS = listOf(
    SessionBadgeType.RACE,
    SessionBadgeType.QUALY,
    SessionBadgeType.SPRINT,
    SessionBadgeType.PRACTICE
)

/** Panel de filtro de Eventos (búsqueda + series + tipo de sesión) — vive detrás del botón
 *  de embudo de la barra superior, ver EventsScreen. Extraído de EventsScreen.kt (Fase 4 del
 *  diagnóstico) sin cambiar ningún comportamiento, solo para que ese archivo deje de
 *  concentrar navegación + filtros + lista + diálogos en un único sitio de 872 líneas. */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun EventsFilterPanel(
    visible: Boolean,
    searchQuery: String,
    onSearchQueryChange: (String) -> Unit,
    selectedSeries: Set<RaceSeries>,
    onSeriesChange: (Set<RaceSeries>) -> Unit,
    seriesConfigByKey: Map<RaceSeries, SeriesConfigEntity>,
    selectedSessionTypes: Set<String>,
    onSessionTypesChange: (Set<String>) -> Unit
) {
    // Antes esto se ocultaba también con "&& !noEventsAtAll": si un filtro dejaba la lista a
    // cero eventos, el panel con los chips de serie —la única forma de QUITAR ese filtro sin
    // salir de la pantalla— desaparecía con la lista, dejando al botón de embudo sin ningún
    // efecto visible. El panel ahora se ve siempre que se pida, haya o no eventos que mostrar
    // debajo.
    AnimatedVisibility(
        visible = visible,
        enter = expandVertically() + fadeIn(),
        exit = shrinkVertically() + fadeOut()
    ) {
        Column(modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)) {
            OutlinedTextField(
                value = searchQuery,
                onValueChange = onSearchQueryChange,
                modifier = Modifier.fillMaxWidth(),
                placeholder = { Text(tr("events_search_placeholder")) },
                leadingIcon = { Icon(Icons.Default.Search, contentDescription = null) },
                trailingIcon = {
                    if (searchQuery.isNotEmpty()) {
                        IconButton(onClick = { onSearchQueryChange("") }) {
                            Icon(Icons.Default.Close, contentDescription = tr("events_clear_search"))
                        }
                    }
                },
                singleLine = true,
                shape = MaterialTheme.shapes.small
            )

            FlowRow(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 10.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                FilterChip(
                    selected = selectedSeries.isEmpty(),
                    onClick = { onSeriesChange(emptySet()) },
                    label = { Text(tr("events_filter_all_series")) }
                )
                RaceSeries.entries.forEach { series ->
                    val active = series in selectedSeries
                    FilterChip(
                        selected = active,
                        onClick = {
                            onSeriesChange(if (active) selectedSeries - series else selectedSeries + series)
                        },
                        label = { Text(seriesConfigByKey[series]?.tag ?: series.defaultTag) }
                    )
                }
            }

            FlowRow(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 10.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                FilterChip(
                    selected = selectedSessionTypes.isEmpty(),
                    onClick = { onSessionTypesChange(emptySet()) },
                    label = { Text(tr("events_filter_all_sessions")) }
                )
                SESSION_TYPE_FILTER_OPTIONS.forEach { badge ->
                    val active = badge in selectedSessionTypes
                    FilterChip(
                        selected = active,
                        onClick = {
                            onSessionTypesChange(if (active) selectedSessionTypes - badge else selectedSessionTypes + badge)
                        },
                        label = { Text(tr(SessionBadgeType.labelKey(badge))) }
                    )
                }
            }
        }
    }
}
